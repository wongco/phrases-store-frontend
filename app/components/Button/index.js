/* eslint-disable indent */
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Button = styled.button`
  width: 5em;
  border: 1px solid ${({ borderColor }) => borderColor || 'black'};
  color: ${({ color, disabled }) => (!disabled && color) || 'gray'};
  background-color: ${({ bgColor, disabled }) => !disabled && bgColor};

  &:hover {
    background-color: ${({ hoverBgColor, disabled }) =>
      !disabled && hoverBgColor};
    color: ${({ hoverColor, disabled }) => !disabled && hoverColor};
  }
`;

Button.propTypes = {
  disabled: PropTypes.bool,
  color: PropTypes.string,
  borderColor: PropTypes.string,
  hoverBgColor: PropTypes.string,
  hoverColor: PropTypes.string,
};

export default Button;
