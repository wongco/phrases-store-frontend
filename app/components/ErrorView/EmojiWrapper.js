import styled from 'styled-components';

const EmojiWrapper = styled.div`
  font-size: 150%;
  margin: 0 0.5em;
`;

export default EmojiWrapper;
