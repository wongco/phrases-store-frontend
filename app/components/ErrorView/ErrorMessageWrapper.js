import styled from 'styled-components';

const ErrorMessageWrapper = styled.div`
  margin: 0.5em;
`;

export default ErrorMessageWrapper;
