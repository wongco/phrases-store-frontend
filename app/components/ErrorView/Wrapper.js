import styled from 'styled-components';

const Wrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  margin: 1em 0;
`;

export default Wrapper;
