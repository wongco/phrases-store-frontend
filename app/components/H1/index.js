import styled from 'styled-components';

const H1 = styled.h1`
  text-align: center;
  font-size: 200%;
  font-weight: 200;
`;

export default H1;
