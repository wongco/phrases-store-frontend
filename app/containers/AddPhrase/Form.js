import styled from 'styled-components';

const Form = styled.form`
  display: flex;
  flex-direction: row;
  justify-content: center;
  width: 80%;
  margin-bottom: 1em;
`;

export default Form;
