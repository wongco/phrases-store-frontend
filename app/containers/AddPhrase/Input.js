import styled from 'styled-components';

const Input = styled.input`
  height: 2em;
  padding: 0 1em;
  width: 100%;
  border: 1px solid orange;
`;

export default Input;
