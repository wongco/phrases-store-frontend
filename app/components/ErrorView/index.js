import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Wrapper from './Wrapper';
import EmojiWrapper from './EmojiWrapper';
import ErrorMessageWrapper from './ErrorMessageWrapper';

class ErrorView extends PureComponent {
  static Image = props => <EmojiWrapper>{props.emoji}</EmojiWrapper>;

  static Message = props => (
    <ErrorMessageWrapper>{props.children}</ErrorMessageWrapper>
  );

  render() {
    return <Wrapper>{this.props.children}</Wrapper>;
  }
}

ErrorView.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
};

export default ErrorView;
