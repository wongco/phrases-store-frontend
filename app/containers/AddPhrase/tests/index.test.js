import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import Button from 'components/Button';
import H1 from 'components/H1';
import ErrorView from 'components/ErrorView';
import LoadingIndicator from 'components/LoadingIndicator';
import { AddPhrase } from '../index';
import Form from '../Form';
import Input from '../Input';

describe('<AddPhrase>', () => {
  it('renders without crashing', () => {
    shallow(<AddPhrase />);
  });

  it('matches snapshot for default criteria', () => {
    const wrapper = shallow(<AddPhrase />);
    const serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });

  describe('rendering tests', () => {
    it('renders imported components normally on load', () => {
      const wrapper = shallow(<AddPhrase />);

      const H1Component = wrapper.find(H1);
      expect(H1Component).toHaveLength(1);
      const formComponent = wrapper.find(Form);
      expect(formComponent).toHaveLength(1);
      const inputComponent = wrapper.find(Input);
      expect(inputComponent).toHaveLength(1);
      const buttonComponent = wrapper.find(Button);
      expect(buttonComponent).toHaveLength(1);
    });

    it('should not render ErrorView or LoadingIndicator initially', () => {
      const wrapper = shallow(<AddPhrase loading={false} error={false} />);

      const loadingComponent = wrapper.find(LoadingIndicator);
      expect(loadingComponent).toHaveLength(0);
      const errorView = wrapper.find(ErrorView);
      expect(errorView).toHaveLength(0);
    });

    it('when loading is true, renders LoadingIndicator', () => {
      const wrapper = shallow(<AddPhrase loading />);
      const loadingComponent = wrapper.find(LoadingIndicator);
      expect(loadingComponent).toHaveLength(1);
    });

    it('when error is true, renders ErrorView', () => {
      const wrapper = shallow(<AddPhrase error />);
      const errorView = wrapper.find(ErrorView);
      expect(errorView).toHaveLength(1);
    });
  });

  describe('lifecycle method tests', () => {
    it('upon unmount, clearAppError is called', () => {
      const mockClearAppError = jest.fn();
      const wrapper = shallow(<AddPhrase clearAppError={mockClearAppError} />);
      wrapper.unmount();

      expect(mockClearAppError).toHaveBeenCalled();
    });
  });

  describe('component method tests', () => {
    const mockInputEvent = {
      target: {
        name: 'phraseInput',
        value: 'newTest',
      },
    };

    it('handleChange should update state with new event input value', () => {
      const wrapper = shallow(<AddPhrase />);
      const initialState = { ...wrapper.state() };

      wrapper.instance().handleChange(mockInputEvent);
      const updatedState = { ...wrapper.state() };

      expect(initialState).toEqual({ phraseInput: '' });
      expect(updatedState).toEqual({ phraseInput: 'newTest' });
      expect(updatedState).not.toEqual(initialState);
    });

    it('handleSubmit should invoke preventDefault and dispatch addPhrase', () => {
      const mockAddPhrase = jest.fn();
      const mockPreventDefault = jest.fn();
      const mockSubmitEvent = {
        preventDefault: mockPreventDefault,
      };

      const wrapper = shallow(<AddPhrase addPhrase={mockAddPhrase} />);
      wrapper.instance().handleChange(mockInputEvent);
      wrapper.instance().handleSubmit(mockSubmitEvent);

      expect(mockPreventDefault).toHaveBeenCalled();
      expect(mockAddPhrase).toHaveBeenCalled();
      const textArgument = mockAddPhrase.mock.calls[0][0];
      expect(textArgument).toBe('newTest');
    });
  });
});
