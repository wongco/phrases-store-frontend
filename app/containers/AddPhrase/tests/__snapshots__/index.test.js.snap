// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPhrase> matches snapshot for default criteria 1`] = `
<AddPhraseWrapper>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Add a phrase
    </title>
    <meta
      content="Description of AddPhrase"
      name="description"
    />
  </HelmetWrapper>
  <H1>
    <FormattedMessage
      defaultMessage="Add a phrase time!"
      id="app.containers.AddPhrase.header"
      values={Object {}}
    />
  </H1>
  <Form
    onSubmit={[Function]}
  >
    <Input
      borderColor="orange"
      name="phraseInput"
      onChange={[Function]}
      placeholder="sample phrase: the cow jumped over the moon!"
      type="text"
      value=""
    />
    <Button
      bgColor="#ffe6b3"
      borderColor="gray"
      color="black"
      disabled={true}
      hoverBgColor="orange"
      hoverColor="white"
    >
      Add
    </Button>
  </Form>
</AddPhraseWrapper>
`;
